src/mo_idl/idl_to_mo.ml

```diff
@@ -49,65 +48,73 @@ let is_tuple fs =
          let res_t = M.Con (con, []) in
-         m_env := M.Env.add id res_t !m_env;
+         occs := M.Env.add id res_t !occs;
          let t' = I.Env.find id env in
-         let t' = check_typ env t' in
+         let t' = check_typ' env occs t' in
          M.set_kind con (M.Def ([], t'));
          res_t
       | Some t -> t
      )
-  | OptT t -> M.Opt (check_typ env t)
-  | VecT t -> M.Array (check_typ env t)
+  | OptT t -> M.Opt (check_typ' env occs t)
+  | VecT t -> M.Array (check_typ' env occs t)
   | BlobT -> M.Prim M.Blob
   | RecordT fs ->
      if is_tuple fs then
-       M.Tup (List.map (fun f -> check_typ env f.it.typ) fs)
+       M.Tup (List.map (fun f -> check_typ' env occs f.it.typ) fs)
      else
-       let fs = List.map (check_field env) fs in
+       let fs = List.map (check_field env occs) fs in
        M.Obj (M.Object, List.sort M.compare_field fs)
   | VariantT fs ->
-     let fs = List.map (check_variant_field env) fs in
+     let fs = List.map (check_variant_field env occs) fs in
      M.Variant (List.sort M.compare_field fs)
   | FuncT (ms, ts1, ts2) ->
      let (s, c) = check_modes ms in
-     M.Func (M.Shared s, c, [M.scope_bind], check_typs env ts1, check_typs env ts2)
+     M.Func (M.Shared s, c, [M.scope_bind], check_typs' env occs ts1, check_typs' env occs ts2)
   | ServT ms ->
-     let fs = List.map (check_meth env) ms in
+     let fs = List.map (check_meth env occs) ms in
      M.Obj (M.Actor, List.sort M.compare_field fs)
   | ClassT _ -> raise (UnsupportedCandidFeature
      (Diag.error_message t.at "M0162" "import" "Candid service constructor type not supported as Motoko type"))
   | PreT -> assert false
-and check_typs env ts = List.map (check_typ env) ts
-and check_field env f =
-  M.{lab = check_label f.it.label; typ = check_typ env f.it.typ; depr = None}
-and check_variant_field env f =
+and check_typs' env occs ts = List.map (check_typ' env occs) ts
+and check_field env occs f =
+  M.{lab = check_label f.it.label; typ = check_typ' env occs f.it.typ; depr = None}
+and check_variant_field env occs f =
   match f.it.typ.it with
   | PrimT Null -> M.{lab = check_label f.it.label; typ = M.Tup []; depr = None}
-  | _ -> check_field env f
-and check_meth env (m: typ_meth) =
-  M.{lab = Idllib.Escape.escape_method m.it.var.at m.it.var.it; typ = check_typ env m.it.meth; depr = None}
+  | _ -> check_field env occs f
+and check_meth env occs (m: typ_meth) =
+  M.{lab = Idllib.Escape.escape_method m.it.var.at m.it.var.it; typ = check_typ' env occs m.it.meth; depr = None}
 
 let check_prog (env: typ I.Env.t) actor : M.typ =
+  let occs = ref M.Env.empty in
   match actor with
   | Some {it=ServT ms; _} ->
-     let fs = List.map (check_meth env) ms in
+     let fs = List.map (check_meth env occs) ms in
+     (* TODO: why do we only check and include the mentioned types (occs),
+        and not all of the .did declared ones (available to the caller, if not here? *)
```

Comment:
I don't have a preference here, but if the service didn't mention this type, why does the user need to use that type? `did` file is kind of transparent to the user here: it's stored inside `.dfx` directory, not part of the source code.